### PR TITLE
dispose the ShaderWarmUp picture/image

### PR DESCRIPTION
**Pull request by dnfield**
Disposes of native resources created by `ShaderWarmUp.execute`.

Does this by adding some new debug instrumentation to that method so we can capture its internals for assertions in a test.

Previously attempted here: https://github.com/flutter/flutter/pull/84669

I thought I'd have to write a benchmark for this but .. I do not.

### CODE REVIEW

File: packages/flutter/lib/src/painting/shader_warm_up.dart [hunk 1]
```diff
@@ -85,15 +87,19 @@ abstract class ShaderWarmUp {
     final ui.Canvas canvas = ui.Canvas(recorder);
     await warmUpOnCanvas(canvas);
     final ui.Picture picture = recorder.endRecording();
+    assert(debugCaptureShaderWarmUpPicture(picture));
     if (!kIsWeb) { // Picture.toImage is not yet implemented on the web.
```
Comment (dnfield) [inline packages/flutter/lib/src/painting/shader_warm_up.dart:91]: Is this still true?

File: packages/flutter/lib/src/painting/debug.dart [hunk 2]
```diff
@@ -195,3 +195,27 @@ bool debugAssertAllPaintingVarsUnset(String reason, { bool debugDisableShadowsOv
   }());
   return true;
 }
+
+/// The signature of [debugCaptureShaderWarmUpPicture].
+///
+/// Used by tests to run assertions on the [Picture] created by
+/// [ShaderWarmUp.execute]. The return value indicates whether the assertions
+/// pass or not.
+typedef ShaderWarmUpPictureCapture = bool Function(Picture);
```
Comment (yjbanov) [inline packages/flutter/lib/src/painting/debug.dart:204]: If it's a signature specifically for `debugCaptureShaderWarmUpPicture`, should it be named `DebugShaderWarmUpPictureCapture`?

File: packages/flutter/lib/src/painting/debug.dart [hunk 3]
```diff
@@ -195,3 +195,27 @@ bool debugAssertAllPaintingVarsUnset(String reason, { bool debugDisableShadowsOv
...
+
+/// The signature of [debugCaptureShaderWarmUpImage].
+///
+/// Used by tests to run assertions on the [Image] created by
+/// [ShaderWarmUp.execute]. The return value indicates whether the assertions
+/// pass or not.
+typedef ShaderWarmUpImageCapture = bool Function(Image);
```
Comment (yjbanov) [inline packages/flutter/lib/src/painting/debug.dart:211]: same here

File: packages/flutter/lib/src/painting/debug.dart [hunk 4]
```diff
@@ -195,3 +195,27 @@ bool debugAssertAllPaintingVarsUnset(String reason, { bool debugDisableShadowsOv
...
+
+/// Called by [ShaderWarmUp.execute] immediately after it creates a [Picture].
+///
+/// Tests may use this to capture the picture and run assertions on it.
+ShaderWarmUpPictureCapture debugCaptureShaderWarmUpPicture = (Picture picture) => true;
+
+/// Called by [ShaderWarmUp.execute] immediately after it creates a [Picture].
```
Comment (yjbanov) [inline packages/flutter/lib/src/painting/debug.dart:218]: s/Picture/Image/

File: packages/flutter/lib/src/painting/shader_warm_up.dart (hunk 1 above)
Comment (yjbanov) [reply packages/flutter/lib/src/painting/shader_warm_up.dart:91]: No, but let's make that change in a separate PR? I'm not sure if the HTML renderer covers this sufficiently.

**Review by yjbanov — APPROVED** [review]
lgtm

File: packages/flutter/lib/src/painting/shader_warm_up.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/lib/src/painting/shader_warm_up.dart:91]: the other thing is whether the HTML renderer would even benefit from this. This is meant to force Skia to compile shaders. I'm not sure how/if that does anything on the CanvasKit backend, but it probably does nothing on the HTML...

File: packages/flutter/lib/src/painting/debug.dart (hunk 2 above)
Comment (dnfield) [reply packages/flutter/lib/src/painting/debug.dart:204]: Maybe this would be better as `ShaderWarmUpPictureCallback`.

**Comment by dnfield** [comment]
@Piinks - any idea about the failure here? I can't reproduce it locally and it's only happening on Cirrus... Seems unrelated to this change.

**Comment by dnfield** [comment]
I'm pulling in master and merging it just in case I had bad luck on my commit branch point...

**Comment by Piinks** [comment]
Ok, I'll keep an eye on it. :)

**Comment by dnfield** [comment]
Run passed after merging to master.

**Pull request merged**